File: meme_inator_back/apps/notifications/domain/entities/notification_entity.py
from dataclasses import dataclass
from datetime import datetime
from typing import Optional
from uuid import UUID

from apps.notifications.domain.enums.notification_enum import (
    NotificationType,
    AssociationType,
)


@dataclass
class NotificationEntity:
    notification_id: UUID
    recipient_id: Optional[UUID]
    sender_id: Optional[UUID]

    sender_avatar_url: Optional[str]

    notification_type: NotificationType
    message: str

    is_read: bool
    created_at: datetime

    association_id: Optional[UUID]
    association_type: Optional[AssociationType]

    # -----------------------
    # Domain behaviors
    # -----------------------

    def mark_as_read(self) -> None:
        self.is_read = True

    def is_system_notification(self) -> bool:
        return self.notification_type == NotificationType.SYSTEM_ALERT

    def has_association(self) -> bool:
        return self.association_id is not None and self.association_type is not None

File: meme_inator_front/lib/features/notifications/domain/entities/notification_entity.dart
import 'package:equatable/equatable.dart';
import 'package:meme_inator_front/features/notifications/domain/enums/association_type.dart';
import 'package:meme_inator_front/features/notifications/domain/enums/notification_type.dart';

class NotificationEntity extends Equatable {
  final String notificationId; // UUID as string
  final String? recipientId;
  final String? senderId;
  final String? senderAvatarUrl;
  final NotificationType notificationType;
  final String message;
  final bool isRead;
  final DateTime createdAt;
  final String? associationId;
  final AssociationType? associationType;

  const NotificationEntity({
    required this.notificationId,
    this.recipientId,
    this.senderId,
    this.senderAvatarUrl,
    required this.notificationType,
    required this.message,
    required this.isRead,
    required this.createdAt,
    this.associationId,
    this.associationType,
  });

  NotificationEntity copyWith({bool? isRead}) {
    return NotificationEntity(
      notificationId: notificationId,
      recipientId: recipientId,
      senderId: senderId,
      senderAvatarUrl: senderAvatarUrl,
      notificationType: notificationType,
      message: message,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt,
      associationId: associationId,
      associationType: associationType,
    );
  }

  @override
  List<Object?> get props => [
        notificationId,
        recipientId,
        senderId,
        senderAvatarUrl,
        notificationType,
        message,
        isRead,
        createdAt,
        associationId,
        associationType,
      ];
}